packages/flutter_tools/lib/src/vmservice.dart

```diff
@@ -40,155 +32,543 @@ class VMService {
-    return new CreateDevFSResponse(response.response);
+  void _buildIsolateList() {
+    List<Isolate> isolateList = _isolateCache.values.toList();
+    isolateList.sort(_compareIsolates);
+    isolates.clear();
+    isolates.addAll(isolateList);
   }
 
-  /// List the available file systems.
-  Future<List<String>> listDevFS() {
-    return sendRequest('_listDevFS').then((Response response) {
-      return response.response['fsNames'];
+  void _removeDeadIsolates(List<Isolate> newIsolates) {
+    // Build a set of new isolates.
+    Set<String> newIsolateSet = new Set<String>();
+    newIsolates.forEach((Isolate iso) => newIsolateSet.add(iso.id));
+
+    // Remove any old isolates which no longer exist.
+    List<String> toRemove = <String>[];
+    _isolateCache.forEach((String id, _) {
+      if (!newIsolateSet.contains(id)) {
+        toRemove.add(id);
+      }
     });
+    toRemove.forEach((String id) => _isolateCache.remove(id));
+    _buildIsolateList();
+  }
+
+  @override
+  ServiceObject getFromMap(Map<String, dynamic> map) {
+    if (map == null) {
+      return null;
+    }
+    String type = _stripRef(map['type']);
+    if (type == 'VM') {
+      // Update this VM object.
+      update(map);
+      return this;
+    }
+
+    String mapId = map['id'];
+
+    switch (type) {
+      case 'Isolate': {
+        // Check cache.
+        Isolate isolate = _isolateCache[mapId];
+        if (isolate == null) {
+          // Add new isolate to the cache.
+          isolate = new ServiceObject._fromMap(this, map);
+          _isolateCache[mapId] = isolate;
+          _buildIsolateList();
+
+          // Eagerly load the isolate.
+          isolate.load().catchError((dynamic e, StackTrace stack) {
+            printTrace('Eagerly loading an isolate failed: $e\n$stack');
+          });
+        } else {
+          // Existing isolate, update data.
+          isolate.update(map);
+        }
+        return isolate;
+      }
+      break;
+      case 'FlutterView': {
+        FlutterView view = _viewCache[mapId];
+        if (view == null) {
+          // Add new view to the cache.
+          view = new ServiceObject._fromMap(this, map);
+          _viewCache[mapId] = view;
+        } else {
+          view.update(map);
+        }
+        return view;
+      }
+      break;
+      default:
+        throw new VMServiceObjectLoadError(
+            'VM.getFromMap called for something other than an isolate', map);
+    }
+  }
+
+  // Note that this function does not reload the isolate if it found
+  // in the cache.
+  Future<Isolate> getIsolate(String isolateId) {
+    if (!loaded) {
+      // Trigger a VM load, then get the isolate. Ignore any errors.
+      return load().then((_) => getIsolate(isolateId)).catchError((_) => null);
+    }
+    return new Future<Isolate>.value(_isolateCache[isolateId]);
+  }
+
+  /// Invoke the RPC and return the raw response.
+  Future<Map<String, dynamic>> invokeRpcRaw(
```

Comment:
I like having a wrapper for responses - `Response` as opposed to `Map<String, dynamic>` - it makes method signatures simpler in a lot of places, and you can hang semantic methods off the object; but YMMV.